### ISSUE TITLE
tailscale-login will not authenicate or recognise URL visit

### ISSUE THREAD

**Issue by glennrice** [tailscale/tailscale#52]
Installed the linux client on the latest linux mint but it won't get past waiting for URL visit:

```
AuthURL is https://login2.tails...
sendStatus: authRoutine3: state:url-visit-required
To authenticate, visit:

        https://login2.tailscale.io/a/...

authRoutine: state:url-visit-required
direct.WaitLoginURL
doLogin(regen=false, hasUrl=true)
RegisterReq: onode=[empty] node=[wJVU…+tjE] fup=true
```
Tried in both Firefox and Chromium but same result.

**Comment by crawshaw** [comment]
After visiting the URL and authenticating, do you see a "Authentication Successful" page?

**Comment by glennrice** [comment]
Nope. Just the sign in with e-mail page, which then comes up with the "Sign up with your identity provider":
![image](https://user-images.githubusercontent.com/35488359/74584113-690c3180-501a-11ea-9448-0a081cc3e031.png)

I click and enter my e-mail address, then I get the following:

![image](https://user-images.githubusercontent.com/35488359/74584125-880ac380-501a-11ea-86dc-0d36f9701fe5.png)

I click "Sign up with another identity" as I'm not using Google or Microsoft but that thats me to https://tailscale.com/onboarding page:

![image](https://user-images.githubusercontent.com/35488359/74584148-bdafac80-501a-11ea-96d7-17db6ff55a98.png)

I'll try a windows PC next...

**Comment by glennrice** [comment]
Just checked DNS and requests to login2.tailscale.io and log.tailscale.io aren't being blocked when I run tailscale-login

**Comment by glennrice** [comment]
Same thing happens on a Windows PC

**Comment by glennrice** [comment]
Ok, trying again it doesn't work with e-mail login or Microsoft account login, at least for me. Does the Microsoft account _need_ to be a business account? I thought I'd try it with the home/personal one first.

I have managed to get it logged in with Google account temporarily. This looks like it will let me get things working but e-mail logon would be preferable. Let me know if there's any testing I can help with to get that going. Even the Microsoft account would be preferable to the Google one for me at least.

Happy to help test if needed.

**Comment by bradfitz** [comment]
@glennricem, we just updated our Microsoft oauth2 config to permit non-business/non-school accounts.

**Comment by bradfitz** [comment]
@glennrice, we just updated our Microsoft oauth2 config to permit non-business/non-school accounts.

**Comment by bradfitz** [comment]
As for arbitrary email addresses, that's not something we want to add. See https://tailscale.com/kb/1013/sso-providers which says:

> Can I sign up with an email address?
>
> We don’t support sign-up with email addresses. By design Tailscale is not an identity provider: there are no Tailscale passwords, account recovery, etc.
>
> Using an identity provider is not only more secure than email and password, but it allow us to automatically rotate connection encryption keys, follow security policies set by your team (eg. 2FA), and more.
>
> We plan to support more auth providers in the future. If you’d like to request support for specific providers, please let us know.

See the "identity" label (https://github.com/tailscale/tailscale/labels/identity) for currently open bugs for other auth providers.

**Comment by bradfitz** [comment]
What if we supported logging in via an email address, but no password?

And then to prove they owned that email address, they emailed a token to a magic email address of ours that verified the DKIM signature and parsed the token?

I've written this before (https://github.com/golang/build/blob/442b1a7d81d6d71264b3bf31c305af92fa0cb7dc/cmd/pubsubhelper/pubsubhelper.go#L287) and it was pretty painless.

I've seen a few other sites do email-a-link too (like https://opencollective.com/) as their only login method, but I haven't seen the DKIM thing before.

The UI could be:

* Enter your email: [ brad@danga.com ] [ Generate login email ]
* Email `login@tailscale.com` the token **XYZXYZXYZXYZXYZXYZXYZ** (with a `mailto:login@tailscale.com?subject=XYZXYZXYZXYZXYZXYZXYZ` link)
* ... the page long polls and tells them when it's all good.

**Comment by lpil** [comment]
Magic email links would work great for me, nice idea @bradfitz

**Comment by bradfitz** [comment]
@lpil, to be clear, I'm not talking about magic email links. We would never **send** an email (so no abuse or phishing potential). We would provide you with a single-use email address that you send **to**. We'd then verify the DKIM signature on the email and log you in.

**Comment by DanielHeath** [comment]
+1 for the "do not want to depend on the largess of google / microsoft for my personal security" bucket. DKIM signing is no barrier.

I would also be pretty happy with with "add a DNS record to locate the oauth2 provider for tailscale to use" rather than "contact support to use another provider".

**Comment by 6-AND-9** [comment]
As I said in #2379, if you are aiming at making a truly "**Private** WireGuard® network", using Google/Microsoft/GitHub is directly opposed to it. 

The idea of @bradfitz is very viable. The question is what do we do if one is unable to access their email. Maybe use a 2FA app with time authentication as an alternative?

**Comment by 6-AND-9** [comment]
A list of OS identity provider alternatives: https://solutionsreview.com/identity-management/the-10-best-free-and-open-source-identity-management-tools/

**Comment by 6-AND-9** [comment]
So, any updates on this?

**Comment by DentonGentry** [comment]
I expect that this is a function we will eventually provide, but delivery of this functionality is not imminent. We will post updates on this issue as progress is made.

The set of currently supported identity providers is: https://tailscale.com/kb/1013/sso-providers/
(I understand that none of these are suitable for your use, just adding the link for anyone else browsing this issue)

**Comment by 6-AND-9** [comment]
Thanks on answering, @DentonGentry. It's unfortunate that I can't start using Tailscale until we get actually private identity providers because I really like what you've done here. But my hopes are up for this to be implemented as soon as possible.

Can we help (boost) somehow with this e.g. by donating?

**Comment by DentonGentry** [comment]
We are aware of your interest in the functionality, and we use the level of interest in planning engineering efforts over time.

**Comment by 6-AND-9** [comment]
Here's a good development in that direction, you should take a look at implementing it: https://decentralized-id.com/blockchain/ethereum/

**Comment by leo60228** [comment]
Compared to a standard "username/password + 2FA + password reset" model, the proposed system has the downside that a compromised email would allow bypassing the 2FA requirement for the Tailscale account. This is relatively minor, but maybe if this is implemented it should be possible to additionally require TOTP and/or WebAuthn at the Tailscale level?

**Comment by DentonGentry** [comment]
Since this feature request was filed, several other relevant features have become available:
- Adding to the list of identity providers: GitHub and Apple signin options have been added.
- Fully custom OIDC providers are available, allowing one to bring your own: https://tailscale.com/kb/1240/sso-custom-oidc/

**Comment by geekscrapy** [comment]
+1 for interest for an "offline", "simple" login method

**Comment by DentonGentry** [comment]
Since the last update, support for Passkeys has been added: https://tailscale.com/kb/1269/passkeys/

**Comment by tycooncapital** [comment]
Capitalism in it's purest form. Shame on you tailscale.

**Comment by bradfitz** [comment]
Huh?? Also, wrong spelling of its.

**Comment by tycooncapital** [comment]
@bradfitz fixed the type-o.

**Comment by tycooncapital** [comment]
@bradfitz about the "Huh". Think about it - do you want centralized currency which is connected to your id? ;-)

You forcing people to use Gmail or Microsoft are paving the way to take the small amounts of freedom people still have in this world. 

Please read about critics of capitalism instead of hating me.

You (as the programmer) is responsible for "now" of the now. You are not free of consequences of the future.

**Comment by bradfitz** [comment]
That's why we added passkeys. It's all local. No company controls your passkeys.

**Comment by tycooncapital** [comment]
@bradfitz I appreciate your feedback. However please read following quote about Passkeys:

> (YouTube comment on top video about passkeys)
> "Welcome to the era of the largest vendor lock of all time"

> What bothers me is that there does not appear to be a standard way to export Passkeys or port them between systems at this point. This locks you in and creates a single point of failure (i.e. if you lose access to your Apple/Google account, you lose access your Passkeys for other accounts too). 

> As long as I don't store my registered passkeys on these tech companies, I'm fine. But currently, logging in passwordless with my Google Account using my Yubico Security Key (the blue one with FIDO2) and Google Chrome Browser doesn't work, despite I already turned on that feature on my account. It keeps asking my PIN. I'm on Linux Mint 21.1.

>  They don't support Linux, that's why.
> With this "beginning of the end of the password", I fear the days of Gmail on Linux are numbered. 

Does this sound like a freedom to you?

**Comment by joshtriplett** [comment]
> Since the last update, support for Passkeys has been added: https://tailscale.com/kb/1269/passkeys/

This would partially solve the problem, if it were possible to do initial signup with a passkey. Or, failing that, if it were possible to sign up using an identity provider, create a passkey-only admin, and then delete the identity provider.

**Comment by itpropro** [comment]
@DentonGentry is it still not possible to use tailscale just with plain email and password? A little weird for a security focused product...
